core/match_pattern.py

```diff
@@ -62,44 +62,99 @@ def compute_child_table(suf_tab, lcp_tab):
     return child_table
 
 
-def get_child_intervals(i, j, child_tab: list[Child]) -> list[Tuple[int, int]]:
-    intervals = []
+def is_interval_eligible(a: str, i: int, lcp: int, s: str, suf_tab: list[int]) -> bool:
+    idx_to_hit = suf_tab[i]
+
+    return s[idx_to_hit + lcp] == a
+
+
+def get_lcp(i: int, j: int, child_tab: list[Child], lcp_tab: list[int]):
+    assert i > -1 and j <= len(child_tab)
+
+    # root interval
+    if i == 0 and j == len(child_tab):
+        return 0
+    elif len(child_tab) > j > child_tab[j].up > i:
+        return lcp_tab[child_tab[j].up]
+    else:
+        return lcp_tab[child_tab[i].down]
+
+
+def get_interval(i: int, j: int, a: str, s: str, child_tab: list[Child],
+                 lcp_tab: list[int], suf_tab: list[int]) -> Tuple[int, int]:
     i_1: int = -1
     i_2: int = -1
 
+    assert i > -1 and j <= len(child_tab)
+
+    lcp = get_lcp(i, j, child_tab, lcp_tab)
+
     if i == 0 and j == len(child_tab):
         i_1 = child_tab[i].next_l_index
-    elif (j + 1 < len(child_tab)) and i < child_tab[j + 1].up <= j:
-        i_1 = child_tab[j + 1].up
+    elif len(child_tab) > j > child_tab[j].up > i:
+        i_1 = child_tab[j].up
     else:
         i_1 = child_tab[i].down
 
     # has no children
     if i_1 == -1:
-        return intervals
+        return -1, -1
 
-    intervals.append((i, i_1 - 1))
+    # runs in infinite loop for some test cases, so still returning (lb, rb - 1) instead of (lb, rb)
+    if is_interval_eligible(a, i, lcp, s, suf_tab):
+        return i, i_1 - 1
 
     while child_tab[i_1].next_l_index != -1:
         i_2 = child_tab[i_1].next_l_index
 
-        intervals.append((i_1, i_2 - 1))
+        if is_interval_eligible(a, i_1, lcp, s, suf_tab):
+            return i_1, i_2 - 1
 
         i_1 = i_2
 
-    intervals.append((i_1, j))
+    if is_interval_eligible(a, i_1, lcp, s, suf_tab):
+        return i_1, j
 
-    return intervals
+    # represents undefined
+    return -1, -1
 
 
-def get_lcp(i, j, child_table: list[Child], lcp_tab):
-    if (i < child_table[j + 1].up) and (child_table[j + 1].up <= j):
-        return lcp_tab[child_table[j + 1].up]
-    else:
-        return lcp_tab[child_table[i].down]
+def match_pattern(s: str, p: str, child_tab: list[Child], lcp_tab: list[int], suf_tab: list[int]) -> bool:
+    # considering empty pattern to be a valid string; returns false if the string s is non-empty
+    if len(p) == 0:
+        # len(s) == 1 is empty because we consider sentinel char $ at the end
+        return len(s) == 1
+
+    c: int = 0
+    pattern_found: bool = True
+    (n, m) = (len(s), len(p))
```

Comment:
You don't need parentheses to make tuples, so in an expression like

```python
(a, b) = (x, y)
```

it is more common to simply write

```python
a, b, = x, y
```